worldedit-core/src/main/java/com/sk89q/worldedit/command/util/AsyncCommandBuilder.java

```diff
@@ -0,0 +1,171 @@
+        this.consumer = consumer;
+        return this;
+    }
+
+    public AsyncCommandBuilder<T> onFailure(@Nullable Component message, @Nullable ExceptionConverter exceptionConverter) {
+        checkArgument(message != null || exceptionConverter != null, "Can't have null message AND exceptionConverter");
+        this.failureMessage = message;
+        this.exceptionConverter = exceptionConverter;
+        return this;
+    }
+
+    public AsyncCommandBuilder<T> onFailure(@Nullable String message, @Nullable ExceptionConverter exceptionConverter) {
+        checkArgument(message != null || exceptionConverter != null, "Can't have null message AND exceptionConverter");
+        this.failureMessage = message == null ? null : ErrorFormat.wrap(message);
+        this.exceptionConverter = exceptionConverter;
+        return this;
+    }
+
+    public ListenableFuture<T> buildAndExec(ListeningExecutorService executor) {
+        final ListenableFuture<T> future = checkNotNull(executor).submit(buildTask());
+        if (delayMessage != null) {
+            FutureProgressListener.addProgressListener(future, sender, delayMessage);
+        }
+        if (supervisor != null && description != null) {
+            supervisor.monitor(FutureForwardingTask.create(future, description, sender));
+        }
+        return future;
+    }
+
+    private Callable<T> buildTask() {
+        return () -> {
+            T result = null;
+            try {
+                result = callable.call();
+                if (consumer != null) {
+                    consumer.accept(result);
+                }
+                if (successMessage != null) {
+                    sender.print(successMessage);
+                }
+            } catch (Exception orig) {
+                if (exceptionConverter != null) {
+                    Component failure = failureMessage != null ? failureMessage : TextComponent.of("An error occurred");
+                    try {
+                        exceptionConverter.convert(orig);
+                    } catch (CommandException converted) {
+                        Exception targetEx = converted;
+                        // TODO remove this once WG migrates to piston and can use piston exceptions everywhere
+                        if (converted instanceof CommandExecutionException) {
+                            Throwable parentCause = converted;
+                            while ((parentCause = parentCause.getCause()) != null) {
+                                if (parentCause instanceof com.sk89q.minecraft.util.commands.CommandException) {
+                                    targetEx = ((com.sk89q.minecraft.util.commands.CommandException) parentCause);
+                                    break;
+                                }
+                            }
+                        }
+                        String message = targetEx.getMessage() != null ? targetEx.getMessage() : "An unknown error occurred.";
```

Comment:
You should use `getRichMessage()` or whatever I called it, to retrieve the full `Component` message, in the case of a regular `CommandException` for `targetEx`. Otherwise the message might not be very pretty.

Reply:
done. it doesn't matter much atm since WG only throws (2011)CommandExceptions, but it will once it throws (Piston)CommandExceptions.